### ISSUE TITLE
Empty line after control structures fixer

### ISSUE THREAD

**Issue by GrahamCampbell** [PHP-CS-Fixer/PHP-CS-Fixer#1122]
We need this fixer. :)

**Comment by Soullivaneuh** [comment]
In which preset this one should be added? Symfony? cc @javiereguiluz

**Comment by keradus** [comment]
Symfony is not fully follow that rule.

like
```php
$foo = ...;
if ($foo) {
    ...
}
```

**Comment by Soullivaneuh** [comment]
@keradus If I well understand, with this code:

```php
$foo = ...;
if ($foo) {
    ...
}
$bar = $foo;
```

With Symfony CS, it should be:

```php
$foo = ...;
if ($foo) {
    ...
}

$bar = $foo;
```

With the fixer asked by @GrahamCampbell, is should be:

```php
$foo = ...;

if ($foo) {
    ...
}

$bar = $foo;
```

Right?

Well, in this case, why not two fixers?

* `empty_line_after_control_structure` -> On Symfony
* `empty_line_before_control_structure` -> On Contrib

What do you think?

**Comment by SpacePossum** [comment]
one with config, but only after the current issue/PR count is down ^_^

**Comment by SpacePossum** [comment]
Having two fixers on 1.x and one on master will be a merge hell.

Having one fixer is more logical to me, it prevents code duplication, having less code to write, debug, maintain and document, 
The diff with those three fixers is those have diff. logic to find and test candidate tokens to fix;
namespace, open tag and return.
The new fixers would not, both fixers would have the same logic to find candidates; control structure.
Btw. refactor of `blank_line_before_return` is already proposed to supported more cases using configuration on master.

**Comment by Soullivaneuh** [comment]
> Having two fixers on 1.x and one on master will be a merge hell.

Agree. :+1: 

Keeping two fixers on master could be great too IMHO.

> Having one fixer is more logical to me, it prevents code duplication, having less code to write, debug, maintain and document, 
> The diff with those three fixers is those have diff. logic to find and test candidate tokens to fix;
namespace, open tag and return.

Why not an abstract fixer class for common code? This is already done.

**Comment by SpacePossum** [comment]
>Why not an abstract fixer class for common code? This is already done.

you would still analyze to find the same candidates twice for each file, slowing down the app

**Comment by Soullivaneuh** [comment]
> you would still analyze to find the same candidates twice for each file, slowing down the app

Oh, I see your concern.

**Comment by keradus** [comment]
Ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1122#issuecomment-226136058

With Symfony, both ways are accepted:
```php
$foo = ...;
if ($foo) {
    ...
}

$bar = $foo;
```

```php
$foo = ...;

if ($foo) {
    ...
}

$bar = $foo;
```

**Comment by Soullivaneuh** [comment]
> (so none is enforced)

Oh so both way should be on contrib, not Symfony. right?

**Comment by GrahamCampbell** [comment]
Symfony might want to enforce one of them though. We should ask.

**Comment by keradus** [comment]
fix @GrahamCampbell

**Comment by SpacePossum** [comment]
I've changed the title sine we've `BlankLineBeforeStatementFixer` which can be used to configure to make sure there is a white line before a control structure, so what is left is a way to get a white line after a control structure.

**Comment by shakaran** [comment]
@GrahamCampbell @keradus @SpacePossum @soullivaneuh 

This is the oldest issue in the github tracker for this project. Any remaining or missing here waiting to close?

**Comment by keradus** [comment]
i don't think it;s implemented.
do you wish to implement it?

**Comment by shakaran** [comment]
> i don't think it;s implemented.
> do you wish to implement it?

I would like know better the codebase. Maybe it is still a bit hard to me achieve this, but I can try for learn if someone guides me along the process.

As far I see BlankLineAfterStatementFixer and BlankLineAfterOpeningTagFixer should be merged in a new file called BlankLineAfterStatementFixer.

So I made the pull request https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5888 with that initial change putting under the src/Fixer/Whitespace.

I never code a fixer, so I try to do my best merging and it is unfinished, but I don't know still how to apply the fix for a tag or other. So any feedback or even commit with improvement is welcome.

Also the test are missing.

**Comment by Wirone** [comment]
I am going to close this one because:

- it does not provide clear definition of done (code snippets, expectations)
- it's been 8 years without implementation, so it looks like it's not needed that much 😉 
- I believe most of the cases can be covered with `blank_line_before_statement` (even example from [this comment](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/1122#issuecomment-226136058), with blank line before `if`)

It can be reopened if there's interest, but I believe proper description is required. Right now it's too wide and scope is unknown.

**Comment by Wirone** [comment]
@localheinz Are you planning an implementation 😉? Because if not, I would rather move this to Discussions and try to figure out the definition of done and implementation assumptions. This issue is not described well and discussion is long and hard to follow. Not much value IMHO

**Comment by Wirone** [comment]
Just found there's [`ErickSkrauch/line_break_after_statements`](https://github.com/erickskrauch/php-cs-fixer-custom-fixers#erickskrauchline_break_after_statements), seems like a good alternative 🙂.

**Comment by Wirone** [comment]
Another alternative: [`PedroTroller/line_break_between_statements`](https://github.com/PedroTroller/PhpCSFixer-Custom-Fixers#pedrotrollerline_break_between_statements) 😁.
